File: howto-dapp/src/main/java/de/timolia/howto/generator/PageResponsibilities.java
package de.timolia.howto.generator;

import de.timolia.howto.LanguageUtil;
import de.timolia.howto.Utils;
import de.timolia.howto.models.Language;
import de.timolia.howto.models.Teamler;

import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class PageResponsibilities {

    public static String generate(List<Teamler> teamlers) {

        LinkedHashMap<String, LinkedHashMap<String, String>> responsibilities = new LinkedHashMap<>() {{
            put("%page.responsibilities.project-team-orga%", new LinkedHashMap<>() {{
                put("%page.responsibilities.teamleader%", "%page.responsibilities.teamleader.desc%");
                put("%page.responsibilities.project-team-orga.apply.builder%", null);
                put("%page.responsibilities.project-team-orga.apply.developer%", null);
                put("%page.responsibilities.project-team-orga.apply.supporter%", "%page.responsibilities.project-team-orga.apply.supporter.desc%");
                put("%page.responsibilities.project-team-orga.howto%", "%page.responsibilities.project-team-orga.howto.desc%");
                put("%page.responsibilities.project-team-orga.tournament%", "%page.responsibilities.project-team-orga.tournament.desc%");
                put("%page.responsibilities.project-team-orga.uhc%", "%page.responsibilities.project-team-orga.uhc.desc%");
                put("%page.responsibilities.project-team-orga.youtuber%", "%page.responsibilities.project-team-orga.youtuber.desc%");
            }});
            put("%page.responsibilities.communication-platforms%", new LinkedHashMap<>() {{
                put("%page.responsibilities.communication-platforms.community-discord%", "%page.responsibilities.communication-platforms.community-discord.desc%");
                put("%page.responsibilities.communication-platforms.bug-reports%", null);
                put("%page.responsibilities.communication-platforms.appeal%", null);
                put("%page.responsibilities.communication-platforms.falsely-banned%", null);
                put("%page.responsibilities.communication-platforms.socialmedia%", "%page.responsibilities.communication-platforms.socialmedia.desc%");
                put("%page.responsibilities.communication-platforms.teamspeak%", "%page.responsibilities.communication-platforms.teamspeak.desc%");
            }});
            put("%page.responsibilities.other-responsibilities%", new LinkedHashMap<>() {{
                put("%page.responsibilities.other-responsibilities.graphicdesign%", null);
                put("%page.responsibilities.other-responsibilities.map-submissions%", null);
                put("%page.responsibilities.other-responsibilities.monthly-winner%", null);
                //put("%page.responsibilities.other-responsibilities.translation%", null);
            }});
        }};

        StringBuilder sb = new StringBuilder();
        sb.append("%page.responsibilities.info%");

        for (Map.Entry<String, LinkedHashMap<String, String>> categories : responsibilities.entrySet()) {
            String categoryName = categories.getKey();
            LinkedHashMap<String, String> subcategories = categories.getValue();

            sb
                    .append("\n")
                    .append("\n").append("# ").append(categoryName);

            for (Map.Entry<String, String> e : subcategories.entrySet()) {
                String name = LanguageUtil.translate(Language.DE, e.getKey());
                String description = e.getValue();

                sb
                        .append("\n")
                        .append("\n").append("### **").append(name).append("**");
                if (description != null) {
                    sb.append("\n").append(description).append("\n");
                }

                List<Teamler> hv = teamlers.stream()
                        .filter(teamler -> teamler.hasResponsibilityMain(name))
                        .sorted(Teamler::compare)
                        .collect(Collectors.toList());
                List<Teamler> nv = teamlers.stream()
                        .filter(teamler -> teamler.hasResponsibilitySecondary(name))
                        .sorted(Teamler::compare)
                        .collect(Collectors.toList());

                String titleCustom = null;

                LinkedList<Utils.KeyValuePair<String, String>> hvCustom = new LinkedList<>();
                LinkedList<Utils.KeyValuePair<String, String>> nvCustom = new LinkedList<>();

                // custom stuff
                if (e.getKey().equals("%page.responsibilities.bug-reports%")) {
                    nvCustom.add(new Utils.KeyValuePair<>("%page.responsibilities.custom.bug-reports.modsanddevs%", null));
                } else if (e.getKey().equals("%page.responsibilities.sm%")) {
                    titleCustom = "Projekte u. Verantwortungsgrad";
                    for (Teamler teamler : teamlers.stream().filter(teamler -> teamler.hasResponsibilityMain("Twitter")).sorted(Teamler::compare).collect(Collectors.toList())) {
                        nvCustom.add(new Utils.KeyValuePair<>("|<span class='" + teamler.getRankCurrent().getCssClass() + "'>" + teamler.getNameForMarkdown() + "</span>", "%page.responsibilities.custom.sm.twitter.hv%"));
                    }
                    for (Teamler teamler : teamlers.stream().filter(teamler -> teamler.hasResponsibilitySecondary("Twitter")).sorted(Teamler::compare).collect(Collectors.toList())) {
                        nvCustom.add(new Utils.KeyValuePair<>("|<span class='" + teamler.getRankCurrent().getCssClass() + "'>" + teamler.getNameForMarkdown() + "</span>", "%page.responsibilities.custom.sm.twitter.nv%"));
                    }
                    for (Teamler teamler : teamlers.stream().filter(teamler -> teamler.hasResponsibilityMain("Bauteam-Twitter")).sorted(Teamler::compare).collect(Collectors.toList())) {
                        nvCustom.add(new Utils.KeyValuePair<>("|<span class='" + teamler.getRankCurrent().getCssClass() + "'>" + teamler.getNameForMarkdown() + "</span>", "%page.responsibilities.custom.sm.twitter-builder.hv%"));
                    }
                    for (Teamler teamler : teamlers.stream().filter(teamler -> teamler.hasResponsibilitySecondary("Bauteam-Twitter")).sorted(Teamler::compare).collect(Collectors.toList())) {
                        nvCustom.add(new Utils.KeyValuePair<>("|<span class='" + teamler.getRankCurrent().getCssClass() + "'>" + teamler.getNameForMarkdown() + "</span>", "%page.responsibilities.custom.sm.twitter-builder.nv%"));
                    }
                } else if (e.getKey().equals("%page.responsibilities.teamleader%")) {
                    titleCustom = "%page.responsibilities.custom.teamleader.title%";
                    // |<span class='management'>J4mPr0</span>    | Management des gesamten Timolia-Teams |
                    // |<span class='headbuilder'>Jukplays</span>  | Management des Bau-Teams |
                    for (Teamler teamler : teamlers.stream().filter(teamler -> teamler.hasResponsibilityMain("Management des Timolia-Teams")).sorted(Teamler::compare).collect(Collectors.toList())) {
                        hvCustom.add(new Utils.KeyValuePair<>("|<span class='" + teamler.getRankCurrent().getCssClass() + "'>" + teamler.getNameForMarkdown() + "</span>", "%page.responsibilities.custom.teamleader.whole-team%"));
                    }
                    for (Teamler teamler : teamlers.stream().filter(teamler -> teamler.hasResponsibilityMain("Management des Bau-Teams")).sorted(Teamler::compare).collect(Collectors.toList())) {
                        hvCustom.add(new Utils.KeyValuePair<>("|<span class='" + teamler.getRankCurrent().getCssClass() + "'>" + teamler.getNameForMarkdown() + "</span>", "%page.responsibilities.custom.teamleader.builder%"));
                    }
                } else if (e.getKey().equals("%page.responsibilities.yter%")) {
                    titleCustom = "%page.responsibilities.custom.yter.title%";
                }


                if (titleCustom == null) {
                    sb.append("\n").append("| %page.responsibilities.teamler% | %page.responsibilities.responsib% |");
                } else {
                    sb.append("\n").append("| %page.responsibilities.teamler% | ").append(titleCustom).append(" |");
                }

                sb.append("\n").append("| ------ | ------ |");

                if (hvCustom.isEmpty()) {
                    for (Teamler teamler : hv) {
                        sb.append("\n").append("|<span class='").append(teamler.getRankCurrent().getCssClass()).append("'>").append(teamler.getNameForMarkdown()).append("</span> | %page.responsibilities.hv% |");
                    }
                } else {
                    for (Utils.KeyValuePair<String, String> hvEntry : hvCustom) {
                        String k = hvEntry.getKey();
                        String v = hvEntry.getValue();
                        if (v == null) {
                            sb.append("\n").append(k).append(" | %page.responsibilities.hv% |");
                        } else {
                            sb.append("\n").append(k).append(" | ").append(v).append(" |");
                        }
                    }
                }

                if (!nv.isEmpty() || !nvCustom.isEmpty()) {
                    sb.append("\n").append("| | |");
                    if (nvCustom.isEmpty()) {
                        for (Teamler teamler : nv) {
                            sb.append("\n").append("|<span class='").append(teamler.getRankCurrent().getCssClass()).append("'>").append(teamler.getNameForMarkdown()).append("</span> | %page.responsibilities.nv% |");
                        }
                    } else {
                        for (Utils.KeyValuePair<String, String> nvEntry : nvCustom) {
                            String k = nvEntry.getKey();
                            String v = nvEntry.getValue();
                            if (v == null) {
                                sb.append("\n").append("|").append(k).append(" | %page.responsibilities.nv% |");
                            } else {
                                sb.append("\n").append(k).append(" | ").append(v).append(" |");
                            }
                        }
                    }
                }


            }
        }

        sb.append("\n");

        return sb.toString();
    }

} 

File: howto-dapp/src/main/java/de/timolia/howto/generator/PageResponsibilities.kt
package de.timolia.howto.generator

import de.timolia.howto.Dapp
import de.timolia.howto.teamler.Teamler
import de.timolia.howto.responsibility.Responsibility
import de.timolia.howto.responsibility.ResponsibilityType
import de.timolia.howto.translate.Language
import java.util.*

object PageResponsibilities {
    fun generate(teamlers: List<Teamler>): String {
        val responsibilities: LinkedHashMap<String, LinkedHashMap<String, String?>> = object : LinkedHashMap<String, LinkedHashMap<String, String?>>() {
            init {
                put("%page.responsibilities.project-team-orga%", object : LinkedHashMap<String, String?>() {
                    init {
                        put("%page.responsibilities.teamleader%", "%page.responsibilities.teamleader.desc%")
                        put("%page.responsibilities.project-team-orga.apply.builder%", null)
                        put("%page.responsibilities.project-team-orga.apply.development%", null)
                        put("%page.responsibilities.project-team-orga.apply.support%", "%page.responsibilities.project-team-orga.apply.support.desc%")
                        put("%page.responsibilities.project-team-orga.howto%", "%page.responsibilities.project-team-orga.howto.desc%")
                        put("%page.responsibilities.project-team-orga.tournament%", "%page.responsibilities.project-team-orga.tournament.desc%")
                        //put("%page.responsibilities.project-team-orga.uhc%", "%page.responsibilities.project-team-orga.uhc.desc%");
                        put("%page.responsibilities.project-team-orga.youtuber%", "%page.responsibilities.project-team-orga.youtuber.desc%")
                    }
                })
                put("%page.responsibilities.communication-platforms%", object : LinkedHashMap<String, String?>() {
                    init {
                        put("%page.responsibilities.communication-platforms.community-discord%", "%page.responsibilities.communication-platforms.community-discord.desc%")
                        //put("%page.responsibilities.communication-platforms.bug-reports%", null);
                        put("%page.responsibilities.communication-platforms.appeal%", null)
                        put("%page.responsibilities.communication-platforms.falsely-banned%", null)
                        put("%page.responsibilities.communication-platforms.socialmedia%", "%page.responsibilities.communication-platforms.socialmedia.desc%")
                        put("%page.responsibilities.communication-platforms.teamspeak%", "%page.responsibilities.communication-platforms.teamspeak.desc%")
                    }
                })
                put("%page.responsibilities.other-responsibilities%", object : LinkedHashMap<String, String?>() {
                    init {
                        put("%page.responsibilities.other-responsibilities.graphicdesign%", null)
                        put("%page.responsibilities.other-responsibilities.map-submissions%", null)
                        put("%page.responsibilities.other-responsibilities.monthly-winner%", null)
                        //put("%page.responsibilities.other-responsibilities.translation%", null);
                    }
                })
            }
        }
        val sb = StringBuilder()
        sb.append("%page.responsibilities.info%")
        for ((categoryName, subcategories) in responsibilities) {
            sb
                    .append("\n")
                    .append("\n").append("# ").append(categoryName)
            val nameConverter = Dapp.translate.forLanguage(Language.DE)
            for (e in subcategories.entries) {
                val name = nameConverter.replaceAll(e.key)
                val description = e.value
                sb.append("\n").append("\n").append("### **").append(name).append("**")
                if (description != null) {
                    sb.append("\n").append(description).append("\n")
                }
                val hv = teamlers
                        .filter { teamler -> teamler.hasResponsibilityMain(name) }
                        .sortedWith { obj, o -> obj.compare(o) }
                val nv = teamlers
                        .filter { teamler -> teamler.hasResponsibilitySecondary(name) }
                        .sortedWith { obj, o -> obj.compare(o) }
                var titleCustom: String? = null
                val hvCustom = LinkedList<Responsibility>()
                val nvCustom = LinkedList<Responsibility>()
                when (e.key) {
                    "%page.responsibilities.bug-reports%" -> nvCustom.add(Responsibility.simple(ResponsibilityType.NV, "%page.responsibilities.custom.bug-reports.modsanddevs%"))
                    "%page.responsibilities.sm%" -> {
                        titleCustom = "Projekte u. Verantwortungsgrad"
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilityMain("Twitter") }.sortedWith { obj, o -> obj.compare(o) }) {
                            nvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.sm.twitter.hv%"))
                        }
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilitySecondary("Twitter") }.sortedWith { obj, o -> obj.compare(o) }) {
                            nvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.sm.twitter.nv%"))
                        }
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilityMain("Bauteam-Twitter") }.sortedWith { obj, o -> obj.compare(o) }) {
                            nvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.sm.twitter-builder.hv%"))
                        }
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilitySecondary("Bauteam-Twitter") }.sortedWith { obj, o -> obj.compare(o) }) {
                            nvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.sm.twitter-builder.nv%"))
                        }
                    }

                    "%page.responsibilities.teamleader%" -> {
                        titleCustom = "%page.responsibilities.custom.teamleader.title%"
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilityMain("Management des Timolia-Teams") }.sortedWith { obj, o -> obj.compare(o) }) {
                            hvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.teamleader.whole-team%"))
                        }
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilityMain("Management des Bau-Teams") }.sortedWith { obj, o -> obj.compare(o) }) {
                            hvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.teamleader.builder%"))
                        }
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilityMain("Management des Content-Teams") }.sortedWith { obj, o -> obj.compare(o) }) {
                            hvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.teamleader.content%"))
                        }
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilityMain("Management des Development-Teams") }.sortedWith { obj, o -> obj.compare(o) }) {
                            hvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.teamleader.development%"))
                        }
                        for (teamler in teamlers.filter { teamler -> teamler.hasResponsibilityMain("Management des Support-Teams") }.sortedWith { obj, o -> obj.compare(o) }) {
                            hvCustom.add(Responsibility.concreteTeamster(teamler, "%page.responsibilities.custom.teamleader.support%"))
                        }
                    }

                    "%page.responsibilities.yter%" -> titleCustom = "%page.responsibilities.custom.yter.title%"
                }
                if (titleCustom == null) {
                    sb.append("\n").append("| %page.responsibilities.teamler% | %page.responsibilities.responsib% |")
                } else {
                    sb.append("\n").append("| %page.responsibilities.teamler% | ").append(titleCustom).append(" |")
                }
                sb.append("\n").append("| ------ | ------ |")
                if (hvCustom.isEmpty()) {
                    for (teamler in hv) {
                        hvCustom.add(Responsibility.concreteTeamster(ResponsibilityType.HV, teamler))
                    }
                }
                for (hvEntry in hvCustom) {
                    sb.append("\n")
                    hvEntry.renderTo(sb)
                }
                if (nvCustom.isEmpty()) {
                    for (teamler in nv) {
                        nvCustom.add(Responsibility.concreteTeamster(ResponsibilityType.NV, teamler))
                    }
                }
                for (nvEntry in nvCustom) {
                    sb.append("\n")
                    println(nvEntry)
                    nvEntry.renderTo(sb)
                }
            }
        }
        sb.append("\n")
        return sb.toString()
    }
}
